src/psyclone/dynamo0p3.py

```diff
@@ -2190,38 +2172,316 @@ class DynHaloExchange(HaloExchange):
+            # there is more than one read field in the list.
+            depth_str_list = []
+            for depth_info in new_depth_info_list:
+                depth_str_list.append(self._depth_str(depth_info))
+            return "max("+",".join(depth_str_list)+")"
+
+    def _depth_str(self, depth_info):
+        '''Interal helper method that returns the depth of a halo dependency
+        as a string'''
+        depth_str = ""
+        if depth_info["var_depth"]:
+            depth_str += depth_info["var_depth"]
+            if depth_info["literal_depth"]:
+                depth_str += "+"
+        if depth_info["literal_depth"]:
+            depth_str += str(depth_info["literal_depth"])
+        return depth_str
+
+    def _simplify_depth_list(self, depth_info_list):
+        '''Halo's may have more than one dependency. This method simplifies
+        multiple dependencies to remove duplicates and any obvious
+        redundancy. For example, if one dependency is for depth=1 and
+        another for depth=2 then we do not need the former as it is
+        covered by the latter. Similarly, if we have a depth=extent+1 and
+        another for depth=extent+2 then we do not need the former as
+        it is covered by the latter.'''
+        new_depth_info_list = []
+        for depth_info in depth_info_list:
+            if depth_info["max_depth"]:
+                new_depth_info_list.append({"var_depth": "",
+                                            "literal_depth": 0,
+                                            "max_depth": True})
+                return new_depth_info_list
+
+        literal_only = 0
+        for depth_info in depth_info_list:
+            var_depth = depth_info["var_depth"]
+            literal_depth = depth_info["literal_depth"]
+            match = False
+            for new_depth_info in new_depth_info_list:
+                if new_depth_info["var_depth"] == var_depth and not match:
+                    new_depth_info["literal_depth"] = max(
+                        new_depth_info["literal_depth"], literal_depth)
+                    match = True
+            if not match:
+                new_depth_info_list.append({"var_depth": var_depth,
+                                            "literal_depth": literal_depth,
+                                            "max_depth": False})
+        return new_depth_info_list
+
+    @property
+    def _compute_halo_info(self):
+        '''Dynamically computes all halo dependencies and returns the
+        required halo information (such as halo depth and stencil type) in a
+        list'''
+        read_dependencies = self.field.forward_read_dependencies()
+        if len(read_dependencies) == 0:
+            raise GenerationError(
+                "Internal logic error. There should be at least one read "
+                "dependence for a halo exchange")
+        depth_info_list = []
+        for read_dependency in read_dependencies:
```

Comment:
As we create and then separately populate a HaloAccess object here there is no simple way to turn this into a list comprehension. Note I am still on the fence as to whether HaloAccess should take a field as a constructor which would allow what you suggest.

Reply:
Update, the code has been simplified so now it is implemented as a list comprehension